storage/api/csk_operations.py

```diff
@@ -0,0 +1,172 @@
+
+This sample demonstrates uploading an object while supplying an encryption key,
+retrieving that object's contents, and finally rotating that key to a new value.
+
+This sample is used on this page:
+
+    https://cloud.google.com/storage/docs/json_api/v1/json-api-python-samples
+
+For more information, see the README.md under /storage.
+"""
+
+import argparse
+import filecmp
+import tempfile
+
+from googleapiclient import discovery
+from googleapiclient import http
+
+from oauth2client.client import GoogleCredentials
+
+
+# You can (and should) generate your own encryption key. Here's a good way to
+# accomplish this with Python:
+#   python -c \
+#     'import base64; import os; print(base64.encodestring(os.urandom(32)))'
+# Although these keys are provided here for simplicity, please remember that it
+# is a bad idea to store your encryption keys in your source code.
+ENCRYPTION_KEY = "4RzDI0TeWa9M/nAvYH05qbCskPaSU/CFV5HeCxk0IUA="
+
+# You can use openssl to quicly calculate the hash of any key. Try running this:
+#   openssl base64 -d <<< ENCRYPTION_KEY | openssl dgst -sha256 -binary \
+#     | openssl base64
+KEY_HASH = "aanjNC2nwso8e2FqcWILC3/Tt1YumvIwEj34kr6PRpI="
+
+ANOTHER_ENCRYPTION_KEY = "oevtavYZC+TfGtV86kJBKTeytXAm1s2r3xIqam+QPKM="
+ANOTHER_KEY_HASH = "/gd0N3k3MK0SEDxnUiaswl0FFv6+5PHpo+5KD5SBCeA="
+
+
+def create_service():
+  """Creates the service object for calling the Cloud Storage API."""
+  # Get the application default credentials. When running locally, these are
+  # available after running `gcloud init`. When running on compute
+  # engine, these are available from the environment.
+  credentials = GoogleCredentials.get_application_default()
+
+  # Construct the service object for interacting with the Cloud Storage API -
+  # the 'storage' service, at version 'v1'.
+  # You can browse other available api services and versions here:
+  #     https://developers.google.com/api-client-library/python/apis/
+  return discovery.build("storage", "v1", credentials=credentials)
+
+
+def upload_object(bucket, filename, encryption_key, key_hash):
+  """Uploads an object, specifying a custom encryption key."""
+  service = create_service()
+
+  with open(filename, "rb") as f:
+    request = service.objects().insert(
+        bucket=bucket, name=filename,
+        # You can also just set media_body=filename, but # for the sake of
+        # demonstration, pass in the more generic file handle, which could
+        # very well be a StringIO or similar.
+        media_body=http.MediaIoBaseUpload(f, "application/octet-stream"))
+    request.headers["x-goog-encryption-algorithm"] = "AES256"
+    request.headers["x-goog-encryption-key"] = encryption_key
+    request.headers["x-goog-encryption-key-sha256"] = key_hash
+
+    # Since your request contains encryption keys, it's a good idea to
+    # instruct any caches and proxies not to store your request.
+    request.headers["Cache-Control"] = "no-store"
```

Comment:
After extensive discussion, we decided to update the docs so that they no longer recommend using the `Cache-Control` header, so this should be removed.